aea/mail/base.py

```diff
@@ -971,10 +993,27 @@ def put_message(
         :param to: the recipient of the envelope.
         :param sender: the sender of the envelope.
         :param protocol_id: the protocol id.
-        :param message: the content of the message.
+        :param message: the message.
+        :param context: the envelope context
         :return: None
         """
+        assert isinstance(
+            message, Message
+        ), "Only messages allowed in envelope message field."
+        assert (
+            to or message.counterparty
+        ), "Either provide to or message.counterparty field."
+        assert (
+            to == message.counterparty
+        ), "Inconsistent to and message.counterparty field."
+        assert (
+            protocol_id == message.protocol_id
+        ), "Inconsistent protocol_id and message."
         envelope = Envelope(
-            to=to, sender=sender, protocol_id=protocol_id, message=message
+            to=to or message.counterparty,
+            sender=sender or self._default_address,
+            protocol_id=protocol_id or message.protocol_id,
+            message=message,
+            context=context,
         )
```

Comment:
We now enable the developer to simply do `put_message(message)` which is a large gain in terms of less LOC.